mountinfo/mounted_linux_test.go

```diff
@@ -0,0 +1,254 @@
+	mnt = filepath.Join(dir, "bind-mount-dir")
+	err = os.Mkdir(mnt, 0750)
+	if err != nil {
+		return
+	}
+
+	err = unix.Mount(mnt, mnt, "", unix.MS_BIND, "")
+	if err != nil {
+		err = &os.PathError{Op: "mount", Path: mnt, Err: err}
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// A directory bind-mounted to other directory.
+	src := filepath.Join(dir, "some-dir")
+	err = os.Mkdir(src, 0750)
+	if err != nil {
+		return
+	}
+
+	mnt = filepath.Join(dir, "bind-mounted-dir2")
+	err = os.Mkdir(mnt, 0750)
+	if err != nil {
+		return
+	}
+	err = unix.Mount(src, mnt, "", unix.MS_BIND, "")
+	if err != nil {
+		err = &os.PathError{Op: "mount", Path: mnt, Err: err}
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// A regular file bind-mounted to itself.
+	mnt = filepath.Join(dir, "bind-mount-file")
+	err = ioutil.WriteFile(mnt, []byte(""), 0640)
+	if err != nil {
+		return
+	}
+
+	err = unix.Mount(mnt, mnt, "", unix.MS_BIND, "")
+	if err != nil {
+		err = &os.PathError{Op: "mount", Path: mnt, Err: err}
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// Not mounted socket.
+	sock := filepath.Join(dir, notMounted+".sock")
+	_, err = net.Listen("unix", sock)
+	if err != nil {
+		return
+	}
+	mounts = append(mounts, sock)
+
+	// Bind-mounted socket.
+	mnt = filepath.Join(dir, "bind-mounted-socket")
+	err = ioutil.WriteFile(mnt, []byte(""), 0640)
+	if err != nil {
+		return
+	}
+	err = unix.Mount(sock, mnt, "", unix.MS_BIND, "")
+	if err != nil {
+		err = &os.PathError{Op: "mount", Path: mnt, Err: err}
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// Not mounted directory.
+	mnt = filepath.Join(dir, notMounted+"-dir")
+	err = os.Mkdir(mnt, 0750)
+	if err != nil {
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// Not mounted file.
+	mnt = filepath.Join(dir, notMounted+"-file")
+	err = ioutil.WriteFile(mnt, []byte(""), 0640)
+	if err != nil {
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// A broken not-mounted symlink.
+	symlink := filepath.Join(dir, notMounted+"-broken-symlink")
+	err = unix.Symlink("non-existent-dest", symlink)
+	if err != nil {
+		err = &os.PathError{Op: "symlink", Path: symlink, Err: err}
+		return
+	}
+	mounts = append(mounts, symlink)
+
+	// A valid not-mounted symlink.
+	dst := filepath.Join(dir, "file")
+	err = ioutil.WriteFile(dst, []byte(""), 0640)
+	if err != nil {
+		return
+	}
+	symlink = filepath.Join(dir, notMounted+"-valid-symlink")
+	err = unix.Symlink(dst, symlink)
+	if err != nil {
+		err = &os.PathError{Op: "symlink", Path: symlink, Err: err}
+		return
+	}
+	mounts = append(mounts, symlink)
+
+	// A valid bind-mounted symlink
+	mnt = filepath.Join(dir, "bind-mounted-symlink")
+	err = ioutil.WriteFile(mnt, []byte(""), 0640)
+	if err != nil {
+		return
+	}
+	err = unix.Mount(symlink, mnt, "", unix.MS_BIND, "")
+	if err != nil {
+		err = &os.PathError{Op: "mount", Path: mnt, Err: err}
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	return
+}
+
+func cleanupMounts(t *testing.T, dir string, mounts []string) {
```

Comment:
(not for this PR, but just thinking out loud); we should have a look at the `testing.Cleanup()` option that's now available in Go 1.13+